### ISSUE TITLE
[Enhancement] touchScrollTarget needs even more setup to actually work.

### ISSUE THREAD

**Issue by rockmandash** [jossmac/react-scrolllock#33]
Hello, first, thank you for this wonderful library.
Please see the following code without scrolllock.
```javascript
class App extends Component {
  constructor(props) {
    super(props);
    this.ref = React.createRef();
  }
  render() {
    console.log(this.ref.current); // this will be null, no second render because niether state nor props changed
    return <div ref={this.ref} someprop={this.ref.current} />;  // someprop gets null
  }
}
```
A way to workaround will be like this
```javascript
class App extends Component {
  constructor(props) {
    super(props);
    this.state = { ref: null }
    this.ref = React.createRef();
  }
  componentDidMount() {
    this.setState({ ref: this.ref.current })
  }
  
  render() {
    return <div ref={this.ref} someprop={this.state.ref.current} />;  // someprop gets null first, and actual dom second
  }
}
```
However, this won't work with <Scrolllock /> touchScrollTarget prop
```javascript
class App extends Component {
  constructor(props) {
    super(props);
    this.state = { ref: null }
    this.ref = React.createRef();
  }
  componentDidMount() {
    this.setState({ ref: this.ref.current })
  }
  
  render() {
    return (
      <div>
        <div ref={this.ref} />
        <ScrollLock touchScrollTarget={this.state.ref.current} /> // this is not working
      </div>
    )
  }
}
```
You will have to do this
```javascript
class App extends Component {
  constructor(props) {
    super(props);
    this.state = { ref: null }
    this.ref = React.createRef();
  }
  componentDidMount() {
    this.setState({ ref: this.ref.current })
  }
  
  render() {
    return (
      <div>
        <div ref={this.ref} />
        <ScrollLock key={this.state.ref} touchScrollTarget={this.state.ref.current} /> // add key
      </div>
    )
  }
}
```
add additional key to trigger componentDidMount twice.

**Comment by eugene1g** [comment]
We use react-scrollock with modals, and doing `setState` on the parent container (i.e. the big modal) could be rather expensive because it may trigger re-render of many heavy components inside the modal. I wrote a wrapper around `react-scrollock` that adds the option to provide `touchScrollTarget` as a function instead of an element. If the scroll target element is not immediately available (e.g. because it's rendered at the same time as react-scrolllock), the component will check again on the next tick, then re-render itself using the right element

So your use-case would look like this
```
class App extends Component {  
  render() {
    return (
      <div>
        <div ref={el=>this.scrollEl = el} />
        <EventualScrollLock touchScrollTarget={()=>this.scrollEl} />
      </div>
    )
  }
}
```

And the implementation is -
```js

class EventualScrollLock extends React.Component {
  props: {
    touchScrollTarget?: HTMLElement | (() => HTMLElement | null) | null
  };
  state = {
    actuallyRender: true
  };
  numAttempts = 0;
  scrollableElement;
  refreshRequest;

  constructor(props) {
    super(props);

    if (props.touchScrollTarget instanceof HTMLElement) {
      this.scrollableElement = props.touchScrollTarget;
    }
  }

  componentDidMount() {
    if (typeof this.props.touchScrollTarget === "function") {
      this.tryGetScroller();
    }
  }

  tryGetScroller = () => {
    this.scrollableElement = this.props.touchScrollTarget();
    if (this.scrollableElement) {
      this.setState({ actuallyRender: true });
      return;
    }

    // Safety valve - try to find the scroll element for 120 cycles
    // which is ~2 seconds, then bail to avoid possibly creating an infinite loop
    if (this.numAttempts > 120) {
      this.setState({
        actuallyRender: true
      });
      return;
    }

    // Could not find the scrollable element
    // Wait on rendering until that element is found
    this.setState({
      actuallyRender: false
    });
    // Try to find the element on the next tick
    this.refreshRequest = window.requestAnimationFrame(this.tryGetScroller);
    // Increase the safety valve counter
    this.numAttempts++;
  };

  componentWillUnmount() {
    // Cleanup the timers if this component is unmounted before finding the scroll element
    if (this.refreshRequest && !this.scrollableElement) {
      window.cancelAnimationFrame(this.refreshRequest);
    }
  }

  render() {
    if (!this.state.actuallyRender) return null;
    const { touchScrollTarget, ...restProps } = this.props;
    // Proxy any other props as-is
    return (
      <ScrollLock {...restProps} touchScrollTarget={this.scrollableElement} />
    );
  }
}

```

**Comment by jossmac** [comment]
@rockmandash @eugene1g I've put together a PR to simplify the touch scroll API. Would you mind taking a look at #35?

**Comment by rockmandash** [comment]
@eugene1g Sorry for the late reply!
@jossmac It looks really nice!

**Comment by jossmac** [comment]
@rockmandash would this resolve your issue?

**Comment by rockmandash** [comment]
@jossmac This definitely solve this issue.

**Comment by joeruello** [comment]
For people who ran into this issue, we worked around it by just.
```{ref.current && <ScrollLock touchScrollTarget={ref.current} />}```
willMount will not be called on ScrollLock untill the ref has been set up.

**Comment by jossmac** [comment]
Resolved by #35
